engine/api/v2_workflow_run_craft.go

```diff
@@ -0,0 +1,469 @@
+	if projKey != wref.run.ProjectKey || entityVCS.Name != wref.runVcsServer.Name || entityRepo.Name != wref.runRepo.Name || branch != wref.run.WorkflowRef {
+		entityDB, err = entity.LoadByBranchTypeName(ctx, db, entityRepo.ID, branch, entityType, entityName)
+	} else {
+		entityDB, err = entity.LoadByBranchTypeNameCommit(ctx, db, entityRepo.ID, branch, entityType, entityName, wref.run.WorkflowSha)
+	}
+	if err != nil {
+		if sdk.ErrorIs(err, sdk.ErrNotFound) {
+			return "", &sdk.V2WorkflowRunInfo{WorkflowRunID: wref.run.ID, Level: sdk.WorkflowRunInfoLevelWarning, Message: fmt.Sprintf("obsolete workflow dependency used: %s", name)}, nil
+		}
+		return "", nil, err
+	}
+
+	switch entityType {
+	case sdk.EntityTypeAction:
+		var act sdk.V2Action
+		if err := yaml.Unmarshal([]byte(entityDB.Data), &act); err != nil {
+			return "", nil, err
+		}
+		wref.actionsCache[completePath] = act
+	case sdk.EntityTypeWorkerModel:
+		var wm sdk.V2WorkerModel
+		if err := yaml.Unmarshal([]byte(entityDB.Data), &wm); err != nil {
+			return "", nil, err
+		}
+		wref.workerModelCache[completePath] = wm
+	}
+	return completePath, nil, nil
+}
+
+func (api *API) V2WorkflowRunCraft(ctx context.Context, tick time.Duration) error {
+	ticker := time.NewTicker(tick)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case id := <-api.workflowRunCraftChan:
+			api.GoRoutines.Exec(
+				ctx,
+				"V2WorkflowRunCraftChan-"+id,
+				func(ctx context.Context) {
+					ctx = telemetry.New(ctx, api, "api.V2WorkflowRunCraftFromChan", nil, trace.SpanKindUnspecified)
+					if err := api.craftWorkflowRunV2(ctx, id); err != nil {
+						log.Error(ctx, "V2WorkflowRunCraftChan> error on workflow run %s: %v", id, err)
+					}
+				},
+			)
+		case <-ticker.C:
+			ids, err := workflow_v2.LoadCratingWorkflowRunIDs(api.mustDB())
+			if err != nil {
+				log.Error(ctx, "V2WorkflowRunCraftChan> unable to start tx: %v", err)
+				continue
+			}
+			for _, id := range ids {
+				api.GoRoutines.Exec(
+					ctx,
+					"V2WorkflowRunCraftDB-"+id,
+					func(ctx context.Context) {
+						ctx = telemetry.New(ctx, api, "api.V2WorkflowRunCraftDB", nil, trace.SpanKindUnspecified)
+						if err := api.craftWorkflowRunV2(ctx, id); err != nil {
+							log.Error(ctx, "V2WorkflowRunCraftDB> error on workflow run %s: %v", id, err)
+						}
+					},
+				)
+			}
+		}
+	}
+}
+
+func (api *API) craftWorkflowRunV2(ctx context.Context, id string) error {
+	_, next := telemetry.Span(ctx, "api.craftWorkflowRunV2.lock")
+	lockKey := cache.Key("api:craftWorkflowRunV2", id)
+	b, err := api.Cache.Lock(lockKey, 5*time.Minute, 0, 1)
+	if err != nil {
+		next()
+		return err
+	}
+	if !b {
+		log.Debug(ctx, "api.craftWorkflowRunV2> run %d is locked in cache", id)
+		next()
+		return nil
+	}
+	next()
+	defer func() {
+		_ = api.Cache.Unlock(lockKey)
+	}()
+
+	_, next = telemetry.Span(ctx, "api.craftWorkflowRunV2.LoadRunByID")
+	run, err := workflow_v2.LoadRunByID(ctx, api.mustDB(), id)
```

Comment:
we don't need to Lock for update the run ?
i wonder if we can handle the lock only with the database...

Reply:
Lock is made with the redis.
Crafting is something that take times ( need to retrieve and unmarshal all subs resources )and I don't want to have a long transaction.